compiler/rustc_mir_build/src/thir/pattern/_match.rs

```diff
@@ -2110,7 +2101,12 @@ fn pat_constructor<'tcx>(
             if let Some(int_range) = IntRange::from_const(tcx, param_env, value, pat.span) {
                 Some(IntRange(int_range))
             } else {
-                Some(ConstantValue(value))
+                match value.ty.kind() {
+                    ty::Ref(_, t, _) if t.is_str() => Some(Str(value)),
+                    ty::Float(_) => Some(FloatRange(value, value, RangeEnd::Included)),
+                    // Non structural-match values are opaque.
+                    _ => None,
```

Comment:
`None` here explicitely means a wildcard, so I'm not sure this is correct. Does this behave correctly? I would expect this implies we'd consider a branch that uses an opaque pattern to exhaustively match anything, which is the opposite of how it should behave.
What's an example of such an opaque constant? I'd like to construct a test.

I would instead add an `Opaque` constructor for this case to force us to treat it correctly. (I've also been meaning to add `Wildcard` as a constuctor instead of returning an `Option` here).

Reply:
I remember trying something like this, and could not figure out all the details. If you want to try to add an `Opaque` constructor, it's easy to construct an opaque constant by using a type that doesn't derive `PartialEq` but manually implements it:

```rust
#[derive(PartialEq)]
struct Opaque(i32);

impl Eq for Opaque {}

const FOO: &&Opaque = &&Opaque(42);

match FOO {
    FOO => {},
    // The following should not lint about unreachable
    Opaque(42) => {}
}
```

amusingly the above ICEs on the nightly playground, not sure if that has my PR yet.

Reply:
>  a type that doesn't derive PartialEq but manually implements it:

But this derives it, no manual impl? Or did you mean `Eq`?

Reply:
I think either will do, we require both for structural equality. I was just lazy and didn't want to write the impl when the `Eq` impl is trivial to write

Reply:
Your example ICEs because of the refs, that's been an open issue for a while.
```rust
#[derive(PartialEq)]
struct Opaque(i32);

impl Eq for Opaque {}

const FOO: Opaque = Opaque(42);
match FOO {
    FOO => {},
    // The following should not lint about unreachable
    Opaque(42) => {}
}
```
This works, but does report the second branch as unreachable on both stable and nightly. I don't understand why, and I haven't seen it reported as a bug yet.

Reply:
The good thing is that your incorrect usage of `None` above won't be a regression x)

Reply:
I did an experiment by returning ` Some(NonExhaustive)`, which made patterns like `FOO` above lint that they are unreachable, so that's definitely not it. I'm ot sure what `Constructor::Opaque` would do differently from `NonExhaustive`, but I'm going through all the uses of `NonExhaustive` now to analyze that

Reply:
A simpler way to know what the `Opaque` constructor should do is look at all the places where `ConstantValue` was used but the constant wasn't inspected

Reply:
Actually, an easy way to avoid a regression would be that instead of removing `ConstantValue` altogether like you did, you would rename it to `Opaque` and see what's left.

Reply:
All in all, given that matching with opaque constants is an error anyways, this will not allow any unsound code. So I'm not too worried if we let that slip. We can file the above as a bug and I'll introduce `Opaque` myself whenever I get some time and a powerful enough computer to play with rustc